### DISCUSSION TITLE
Design tokens en responsive gedrag

### DISCUSSION THREAD

**Discussion by jeffreylauwers** [nl-design-system/backlog#346] [Feedback gevraagd]
Dit is de start van een verzameling aan vragen, tips en ideeën en testjes voor design tokens en responsive gedrag. En dan met name hoe dit in Figma is toe te passen.

**Uitgangspunten**

- Code = leidend
- Mobile first
- Uitweking zou in en uit te lezen moeten zijn door Token Studio / Style Dictionary.
- Uitwerking zou niet te complex moeten zijn. Liever geen complexe wiskundige formules, tenzij er een set aan 'sane defaults' is waardoor men niet noodzakelijk in hoeft te duiken.
- Uitwerking zou idealiter gelijkwaardige of snellere/langzamere groei moeten faciliteren.
- Aantal Figma component varianten moet te overzien blijven. Kan Figma variables hierbij helpen?

**Handige linkjes**

https://utopia.fyi/
Online tool om fluid type scales op te zetten.

https://fluid-tokenization.vercel.app/
Online tool om een fluid type scale op te zetten waarbij direct de tokens de kopieren zijn.

[Fluid Design with Tokens - Youtube](https://youtu.be/EbykWCBeqBg?si=2ThnQWN8WMvFGspJ)
Video waarin Marco Krenn van Token Studio uitlegt hoe je fluid design tokens kan inzetten met Token Studio.

[Sam's fancy math equations in Tokens Studio - Substack](https://samiamdesigns.substack.com/p/sams-fancy-math-equations-in-tokens)
Uitgebreid artikel waarbij Sam beschrijft hoe je via Token Studio wiskundige regels kunt gebruiken. Niet zeker in hoeverre ons dit helpt maar ik deel hem hier wel alvast.

**Comment by jeffreylauwers** [comment]
## Eerste uitvraag

Vrij snel na het opzetten van de eerste Figma componenten in combinatie met Token Studio [vroeg ik me voor het eerst af](https://codefornl.slack.com/archives/C02CWGL2M18/p1686124330431639) hoe responsive gedrag hierbij ingebakken kon worden. Dit betrof toen als aanzet enkel een aanpassing in spacing.

Ik kwam toen tot het volgende voorstel. ([Bekijk in Figma Schetsboek](https://www.figma.com/design/fy08SZpZmqx6ljLwvA3Woe/NLDS---Schetsboek?node-id=715-3033&t=45EznIa2kgvcezhh-4))

**Viewport = Small**
`nl.card.padding-inline` = 16px

**Viewport = Medium**
`nl.card.padding-inline` = 32px --> Waarde wijzigt buiten token naamgeving om. Als thema. Net zoals bijvoorbeeld dark-mode.
_of_
`nl.card.vp-medium.padding-inline` = 32px --> ‘Viewport’ (VP) in naamgeving. Een Button kan bijvoorbeeld ook ‘large’ of ‘small’ van zichzelf kan zijn. Zie afbeelding 2.
_of_
`nl.card.{?}.padding-inline` = 32px

<img width="1440" alt="Small en Medium viewport met een Card component. Bij dit Card component verschilt de inline-padding per viewport" src="https://github.com/user-attachments/assets/a9155c53-bed8-446f-a750-0b910347341c">


<img width="1440" alt="Small en Medium viewport met een Button component. Bij dit Button component verschilt de inline-padding per viewport" src="https://github.com/user-attachments/assets/a0c1f6dd-7e86-4516-a0ea-7b48fc32c4af">

Disclaimer: Bovenstaande tokens zijn puur als voorbeeld bedoeld.

**Comment by jeffreylauwers** [comment]
## Voorstel tijdens Design Open Dag september

Tijdens de Design Open Dag van september 2024 zijn we hier met een aantal designers en developers nog eens naar gaan kijken. Op dat moment voelde we ons het meest senang bij de oplossing om de viewport grootte mee te nemen in de token naam. Het idee was om dan een vertaalslag te maken via Style Dictionary zodat dit op de juiste manier in code zou worden weggeschreven.

**Voorbeeld bij een Heading 1 font-size:**
- `nl.heading-1.small-vw.font-size` = 2.5rem
- `nl.heading-1.large-vw.font-size` = 3rem

**Voorbeeld bij een Button inline padding:**
- `nl.button.small-vw.padding-inline` = 16px
- `nl.button.large-vw.padding-inline` = 24px

Disclaimer: Bovenstaande tokens zijn puur als voorbeeld bedoeld.

**Comment by jeffreylauwers** [comment]
## De uitdaging in Figma en Token Studio

**De tokennaam gelijk houden**
Je kunt de tokennaam gelijkhouden en in Token Studio een extra tokenset aanmaken specifiek voor een grotere viewport. Hier kun je vervolgens enkel de tokens in dupliceren waar bij responsive gedrag een verandering optreed. Maar wat betekent dit voor de 'Apply to document' functionaliteit in Token Studio?

**De tokennaam aanvullen met viewport grootte**
Je kunt binnen Token Studio de tokennaam aanvullen met de viewport grootte. Er komt dan binnen dezelfde tokenset een token bij. Het aantal tokens neemt toe. Ook in de lijst met tokens naar de component. Is dit te overweldigend? En hoe is hiermee de aansluiting op Figma Variables?

Hier moeten we nog meer testjes van uitwerken. Wie helpt er mee? 😉

**Comment by RGRMdesign** [comment]
## Voorbeeld toepassing op Paragraph component door tokennaam aan te vullen met vw aanduiding

Om responsive components te gebruiken in NORA heb ik het `utrecht-paragraph` component gekopieerd naar de Lokale bibliotheek van NORA. Zie het [Figma Paragraph Component in de NORA Bibliotheek](https://www.figma.com/design/97dNBzCXJjg3ApxFFrNMlO/Local---NORA---Bibliotheek?node-id=23-190&t=blAbzZIL3pP8Tgra-1)

### Toelichting:

- Het `utrecht.paragraph.font-size` token is vervangen door:
  - `utrecht.paragraph.small-vw.font-size` en
  - `utrecht.paragraph.large-vw.font-size`
- Ik heb voor mijn toepassing de Variant 'Container' hernoemd naar 'Viewport Width'. 

### Voordelen

- Consumers van het Design System kunnen door middel van het wijzigen van de variant op de instance de weergave (Small of Large) aanpassen
- Consumers van het Design System hoeven Tokens Studio niet te gebruiken
- Globale updates (bijvoorbeeld `font-family` of `color`) kunnen met Tokens Studio toegepast worden op het gehele document (Apply to document) zonder dat de verschillende instances voor verschillende Viewport Widths overschreven worden

### Nadelen

- Componenten kunnen veel tokens nodig hebben om het verschil tussen verschillende viewports te ontwerpen, de meest voorkomende zijn denk ik:
  - padding
  - font-size
  - line-height 
- Ook is het gebruikelijk om voor meer dan 2 viewport widths te ontwerpen, vaak Small, Medium, Large. Het toevoegen en onderhouden hiervan kost tijd en energie doordat je tokens moet dupliceren en hernoemen.
- De tokennaam komt niet overeen met hoe die uiteindelijk (idealiter) in css zal bestaan. Het `small-vw` en `large-vw` gedeelte van de token zal daarin niet voorkomen als je deze in verschillende mediaqueries onderbrengt

**Comment by RGRMdesign** [comment]
## Test met het gebruiken van verschillende token sets
Om te ervaren hoe het zou zijn om te werken met verschillende token sets het ik deze test opgezet: [Figma VW Test](https://www.figma.com/design/97dNBzCXJjg3ApxFFrNMlO/Local---NORA---Bibliotheek?node-id=87-412&t=blAbzZIL3pP8Tgra-1).

Beevindingen en ervaringen staan gedocumenteerd in het Figma Design, voor de volledigheid plaats ik ze ook hieronder als screenshot.

<img width="782" alt="image" src="https://github.com/user-attachments/assets/5072bd75-dc66-4ea8-9407-10f5f32b5a1e">

<img width="781" alt="image" src="https://github.com/user-attachments/assets/6fe80622-2053-4dc4-914f-19b5b15f76db">

<img width="782" alt="image" src="https://github.com/user-attachments/assets/e6e0cde9-d234-4fd9-ace9-8b926180078f">

**Comment by RGRMdesign** [comment]
## Layers/Variants locken op een bepaalde combinatie van token sets in Tokens Studio

Een groot nadeel van het gebruik van meerdere token sets om responsive componenten te maken is dat je gauw instances per ongeluk overschrijft wanneer je globale updates doet en deze toepast op het gehele document.

Ik denk dat wanneer je bepaalde layers/variants zou kunnen vastzetten op een bepaalde combinatie van token sets (of thema) dit nadeel grotendeels opgelost is. Hiervoor heb ik een feature request ingediend bij de Tokens Studio Plugin, zie:

https://github.com/tokens-studio/figma-plugin/issues/3178

**Comment by RGRMdesign** [comment]
## Mobile/Small First benadering in tokennaam

Eerder was het idee om zowel `.small-vw.` als `.large-vw.` tokens te hebben. Ik ga nu bij NORA vanuit de Mobile/Small First gedachte de `.small-vw.` tokens weer terug hernoemen naar de originele naam (bijv. `utrecht.paragraph.font-size`). Dit heeft volgens mij de volgende voordelen:

- Small/Mobile First benadering
- De bestaande components, zonder viewport width tokens gaan niet stuk
- Met Style Directory hoeven straks de `.small-vw.` tokens niet in een media-query gezet te worden (komen standaard in `:root`)
- De workflow voor het uitbreiden met een variant voor grotere schermen is makkelijker, doordat je alleen tokens hoeft toe te voegen voor die viewport width zonder dat je de bestaande tokens hoeft te hernoemen

**Comment by RGRMdesign** [comment]
## Viewport specifieke tokens naar een los css bestand

Ik heb voor NORA nu een implementatie gemaakt waarmee tokens met een large-vw aanduiding in de naam naar een apart css bestand worden geschreven.

Zie https://github.com/nl-design-system/themes/pull/772

**Comment by AlineNap** [comment]
## Uitwerking tokensets - tussentijdse update

Ik ben verschillende opties aan het uitwerken om daarbij de voor- en nadelen per gebruikersgroep te onderzoeken. 

### Gebruikersgroepen
Er zijn 3 soorten gebruikers om rekening mee te houden: 
- Groep A: Component bouwer: Jeffrey en Rozerin, de NLDS maintainers
- Groep B: Themabouwer: Rogier, Aline, etc.
- Groep C: Gebruikers van de Figma bibliotheek die designs bouwen met de componenten

### Opties
1. Optie 1: Viewport specifieke tokens per component 
2. Optie 2: Aparte tokensets 
3. Extra: onderzoeken hoe Figma variables kan helpen 
a. Voor optie 1: Switch voor component variants voor gebruikersgroep C
b. Voor optie 2: Collection toevoegen voor de verschillende viewports
c. Kunnen we iets met TS die alleen de common-tokens of spacing als figma variables exporteert? (Is wel ts pro..)

### Uitwerking
Ik zal voor elke optie een bibliotheek (gebruikersgroep A/B) en template uitwerken (werk gebruikersgroep C), om de voor- en nadelen goed in kaart te brengen. Het template bestaat uit de volgende componenten:

- Heading
- Paragraph
- Heading-group
- Action-group (+button)
- Section
- Grid
- Logo

Ik ben met optie 2 begonnen omdat optie 1 al grotendeels is uitgewerkt door Rogier. 

In de json zullen ook andere componenten staan maar kijk met name naar de uitwerking van deze componenten. 

----

## Optie 1: Viewport specifieke tokens per component 

Nadelen die nog niet benoemd zijn voor deze optie:

### Nadeel
1. Gebruikersgroep A/B: als je je font-size aanpast op basis van de viewport heb je van elk Figma component 2 of meerdere variants nodig.
2. Gebruikersgroep C: moet voor elk component de juiste viewport selecteren.
3. Gebruikersgroep B (en C): Als je de font-size niet aanpast per viewpoort heb je onnodige variants (en tokens) in je bibliotheek. 

----

## Optie 2: Aparte tokensets 

### Voordeel
1. De bibliotheek heeft geen onnodige component variants voor thema's die geen responsieve font hebben. 
2. Als je je tokens goed toepast kun je met 1 druk op de knop wisselen van design (gebruikers groep C), zie onderstaande gif. 
3. Ligt dichtbij de code
4. Extra voordeel is dat je met dezelfde opbouw Figma variables kunt inzetten voor degene die dat willen (en daarmee ook meer toekomstbestendig als Figma Variables beter wordt ingericht). Met de opbouw van _'optie 1: viewport specifieke tokens per component_' is dat lastig tot niet mogelijk.

### Nadeel
1. Gebruikersgroep C heeft TS nodig
3. Gebruikersgroep C moet design tokens toepassen voor de lay-out, of via component templates kopiëren/plakken
4. De width, min-width, max-width kunnen niet overschreven worden bij een instance van een component. Als je je logo bijvoorbeeld 40px wilt op mobiel en 48px op desktop kan dat niet door een token toe te passen op de width van het component. Het component luistert naar de width die uit de bibliotheek komt. Alle andere properties/ tokens kun je wel overschrijven: font-size, line-height, padding, gap etc. 
7. Je kunt niet van fill naar hug of andersom. 

Punt 1 t/m 3 kan mogelijk opgelost worden met Figma variables. Zal later uitzoeken hoe en of dat zonder TS PRO en met max 4 Figma variables modes kan.

### Files

- De [Figma bibliotheek](https://www.figma.com/design/xvvZRSM8WNsqZdh56XcnaE/Groep-A?node-id=15622-38066&t=X9OcuD2Ffz1duhR6-1) voor gebruikersgroep A/B
- De [Figma file](https://www.figma.com/design/WjaLGaKBGVeHIGqFEiN8wJ/Groep-C?node-id=0-1&t=7m6PNlePQPz8TRpX-1) voor gebruikersgroep C
- De token.json opzet vind je op [mijn GitHub](https://github.com/AlineNap/LUX/blob/main/responsiveness/aparte-token-sets.json).

### mov

https://github.com/user-attachments/assets/4065e262-ebc8-4a05-9d66-ed6b288a59bc

**Reply by eslook** [reply]
Dit is ongeveer wat wij [nu](https://codefornl.slack.com/archives/C01DAT4TRPF/p1751540405129729) besproken hebben rondom de vraag van Bob voor Open Formulieren in Den Haag! Wij hadden het zo in gedachten:

```
  "viewport/olifant/common": {
    "basis": {
      "color": {
        "transparent": {
          "$type": "color",
          "$value": "transparent"
        }
      }
    }
  },
  "viewport/olifant/components/progress-list/todo": {
    "todo": {
      "progress-list": {
        "step-marker": {
          "current": {
            "background-color": {
              "$type": "color",
              "$value": "{basis.color.accent-1-inverse.bg-default}"
            }
          }
        }
      }
    }
  },
```

en dan ergens dat dan ook de verschillende viewports worden gedefinieerd, dus dat `olifant` bijvoorbeeld translate naar 768px.

Dit is gebaseerd hoe we het (bijna merged) voor darkmode gaan doen:

```
  "color-scheme-dark/common": {
    "basis": {
      "color": {
        "transparent": {
          "$type": "color",
          "$value": "transparent"
        }
      }
    }
  },
  "color-scheme-dark/components/progress-list/todo": {
    "todo": {
      "progress-list": {
        "step-marker": {
          "current": {
            "background-color": {
              "$type": "color",
              "$value": "{basis.color.accent-1-inverse.bg-default}"
            }
          }
        }
      }
    }
  },
  ```

**Comment by jeffreylauwers** [comment]
Brainfart:

Ik krijg steeds meer het gevoel dat het vergroten van tekst zou kunnen plaats vinden in de Brand of Common tokens doormiddel van de clamp functie ([zoals bij Amsterdam](https://github.com/Amsterdam/design-system/blob/develop/proprietary/tokens/src/brand/ams/text.tokens.json)). En dat vanuit de Component tokens simpelweg naar deze token kan worden verwezen.

Responsive gedrag in spacing zie ik dan juist wel weer op Component niveau. En daarbij voorzie ik, in tegenstelling tot fontsizes, juist weer px waardes na het lezen van dit artikel: https://www.joshwcomeau.com/css/surprising-truth-about-pixels-and-accessibility/

**Reply by VincentSmedinga** [reply]
De eerste link is intussen helaas veranderd. Onze brand typography tokens [vind je nu hier](https://github.com/Amsterdam/design-system/blob/develop/packages-proprietary/tokens/src/brand/ams/typography.tokens.json).

**Comment by jeffreylauwers** [comment]
[Fluid Design with Tokens - Youtube](https://youtu.be/EbykWCBeqBg?si=2ThnQWN8WMvFGspJ)
Video waarin Marco Krenn van Token Studio uitlegt hoe je fluid design tokens kan inzetten met Token Studio.

Eerder noteerde ik deze video bij 'Handige linkjes'. Maar dit lijkt, volgens de maker, behoorlijk van de performance te vragen.

![Scherm­afbeelding 2024-12-03 om 12 05 14](https://github.com/user-attachments/assets/ae058969-3658-4960-a768-0578593d982f)

**Comment by jeffreylauwers** [comment]
Voor typografie en responsive gedrag is de volgende Github Discussion in de Rijkshuisstijl community goed om hier te vermelden: [Typografische schaal](https://github.com/nl-design-system/rijkshuisstijl-community/discussions/2072).
